plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Cse.hs

```diff
@@ -0,0 +1,398 @@
+import UntypedPlutusCore.Core
+import UntypedPlutusCore.Purity (isWorkFree)
+import UntypedPlutusCore.Size
+
+import Control.Lens
+import Control.Monad
+import Control.Monad.Trans.Class
+import Control.Monad.Trans.Reader
+import Control.Monad.Trans.State.Strict
+import Data.Foldable
+import Data.Hashable
+import Data.HashMap.Strict (HashMap)
+import Data.HashMap.Strict qualified as Map
+import Data.List.Extra
+import Data.Traversable
+
+{- Note [CSE]
+
+-------------------------------------------------------------------------------
+1. Simplifications
+-------------------------------------------------------------------------------
+
+This is a simplified (i.e., not fully optimal) implementation of CSE. The two simplifications
+we made are:
+
+- No alpha equivalence check, i.e., `\x -> x` and `\y -> y` are considered different expressions.
+- The builtin function arity information is approximate: rather than using the accurate arities,
+  we simply use the maximum number of arguments applied to a builtin function in the program
+  as the builtin function's arity. The arity information is used to determine whether a builtin
+  application is possibly saturated.
+
+-------------------------------------------------------------------------------
+2. How does it work?
+-------------------------------------------------------------------------------
+
+We use the following example to explain how the implementation works:
+
+\x y -> (1+(2+x))
+        +
+        (case y [ (1+(2+x)) + (3+x)
+                , (2+x) + (3+x)
+                , 4+x
+                ]
+        )
+
+The implementation makes several passes on the given term. The first pass collects builtin
+arity information as described above.
+
+In the second pass, we assign a unique ID to each `LamAbs`, `Delay`, and each `Case` branch.
+Then, we annotate each subterm with a path, consisting of IDs encountered from the root
+to that subterm (not including itself).
+
+In the above example, the ID of `\x` is 0, the ID of `\y` is 1, and the IDs of the
+three case branches are 2, 3, 4 (the actual numbers don't matter, as long as they are unique).
+The path for the first `1+(2+x)` and the first `2+x` is "0.1"; the path for the second
+`1+(2+x)` and the second `2+x` is "0.1.2"; the path for `4+x` is "0.1.4".
+
+In the third pass, we calculate a count for each `(term, path)` pair, where `term` is a
+non-workfree term, and `path` is its path. If the same term has two paths, and one is an
+ancestor (i.e., prefix) of the other, we increment the count for the ancestor path in both
+instances.
+
+In the above example, there are three occurrences of `2+x`, whose paths are "0.1", "0.1.2"
+and "0.1.3", respectively. The first path is an ancestor of the latter two. Therefore,
+the count for `(2+x, "0.1")` is 3, while the count for `(2+x, "0.1.2")` and `(2+x, "0.1.3")`
+is 0. The following all have a count of 1: `(3+x, "0.1.2")`, `(3+x, "0.1.3")` and
+`(4+x, "0.1.4")`.
```

Comment:
Because an expression may have multiple occurrences in non-0.1 pairs, e.g., one in 0.1.3 and one in 0.1.3.2.

Reply:
> Because an expression may have multiple occurrences in non-0.1 pairs, e.g., one in 0.1.3 and one in 0.1.3.2.

Sure, but once you have `0.1`, what's even the point of storing other ones? Just increment the counter of `0.1`. Regardless of whether you encounter `0.1.3.2` or `0.1.3.6.3.2.1.6.7.5.8.4` later, it'll still become a variable bound at `0.1`. Is that what you mean by their count be 0? Still haven't reviewed the code.

Reply:
I think this comment makes more sense in the context of your other idea about indexing via path.

Reply:
> but once you have 0.1, what's even the point of storing other ones?

I guess I don't understand your point - but this is exactly what I'm doing, no? If the same expression has "0.1", "0.1.3" and "0.1.3.2", then the count for "0.1" will be incremented in all three cases, and in the end, the count of "0.1" is 3, and the count of the other two paths is 0.

Reply:
Sure, I think Roman's point is just that there's no point even _storing_ the counts for subsumed paths.